timemachine/cpp/src/kernels/k_nonbonded.cuh

```diff
@@ -422,12 +422,165 @@ void __global__ k_nonbonded_unified(
+                int charge_param_idx_j = params_j_idx + PARAM_OFFSET_CHARGE;
+                int lj_param_idx_sig_j = params_j_idx + PARAM_OFFSET_SIG;
+                int lj_param_idx_eps_j = params_j_idx + PARAM_OFFSET_EPS;
+                int w_param_idx_j = params_j_idx + PARAM_OFFSET_W;
+
+                RealType qj = params[charge_param_idx_j];
+                RealType sig_j = params[lj_param_idx_sig_j];
+                RealType eps_j = params[lj_param_idx_eps_j];
+                RealType w_j = params[w_param_idx_j];
+
+                RealType cj_x = coords[atom_j_idx * 3 + 0];
+                RealType cj_y = coords[atom_j_idx * 3 + 1];
+                RealType cj_z = coords[atom_j_idx * 3 + 2];
+
+                RealType delta_x = ci_x - cj_x;
+                RealType delta_y = ci_y - cj_y;
+                RealType delta_z = ci_z - cj_z;
+                RealType delta_w = w_i - w_j;
+
+                delta_x -= bx * nearbyint(delta_x * inv_bx);
+                delta_y -= by * nearbyint(delta_y * inv_by);
+                delta_z -= bz * nearbyint(delta_z * inv_bz);
+
+                RealType d2ij = delta_x * delta_x + delta_y * delta_y + delta_z * delta_z + delta_w * delta_w;
+
+                if (d2ij < cutoff_squared) {
+                    RealType u;
+                    RealType delta_prefactor;
+                    RealType ebd;
+                    RealType dij;
+                    RealType inv_dij;
+                    RealType inv_d2ij;
+                    compute_electrostatics<RealType, true>(
+                        1.0, qi, qj, d2ij, beta, dij, inv_dij, inv_d2ij, ebd, delta_prefactor, u);
+
+                    // lennard jones force
+                    if (eps_i != 0 && eps_j != 0) {
+                        RealType sig_grad;
+                        RealType eps_grad;
+                        compute_lj<RealType, true>(
+                            1.0, eps_i, eps_j, sig_i, sig_j, inv_dij, inv_d2ij, u, delta_prefactor, sig_grad, eps_grad);
+                    }
+                    // Store the atom by atom energy
+                    block_energy_buffer[threadIdx.x] = FLOAT_TO_FIXED_ENERGY<RealType>(u);
+                }
+            }
+            // Sync to ensure the shared buffers are populated
+            __syncthreads();
+
+            block_energy_reduce<THREADS_PER_BLOCK>(block_energy_buffer, threadIdx.x);
+
+            if (threadIdx.x == 0) {
+                output_energies[row_idx * gridDim.x + blockIdx.x] += block_energy_buffer[0];
+            }
+            // Sync the threads so threads don't move on and stomp on the block energy buffer
+            __syncthreads();
+
+            atom_j_idx += gridDim.x * blockDim.x;
+            // Always zero before the end of the loop, in case one thread no longer loops
+            block_energy_buffer[threadIdx.x] = 0;
+        }
+        row_idx += gridDim.y * blockDim.y;
+    }
+}
+
+template <typename RealType, int BLOCK_SIZE>
```

Comment:
`BLOCK_SIZE` == `THREADS_PER_BLOCK`? (If so, best to be consistent in notation)

Reply:
So `BLOCK_SIZE` is the number of threads per block that were run in `k_compute_nonbonded_target_atom_energies` but is not necessarily the number of threads per block run in  `k_compute_nonbonded_target_atom_energies`. 

Will add a comment.

Reply:
Renamed in https://github.com/proteneer/timemachine/pull/1198/commits/50ee91eb83f4c1138b4afd01b10e81bc4acd478d